[TRTLLM-909][feat] Overlap context chunks in pipeline parallel mode

Funatiq opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved batch request tracking and processing to correctly manage context and generation request flows in pipeline execution.
  * Enhanced concurrent request capacity calculations to properly account for distributed processing configuration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

- For context chunks there is no dependency on the results of the last pipeline rank, so they can be scheduled in each iteration.
- To achieve this context requests that are chunking are not added to inflight set, so they are scheduled in the next micro batch.
- Context requests that reach the last context chunk are added to inflight set, so they are not scheduled in the next micro batch and generation can run without overlap.

## Benchmark

Somewhat artificial benchmark to show the benefits:
- Hardware: 2xA100 PCIe with pipeline parallelism (PP=2)
- Model: Qwen3-0.6B
- Dataset: 256 requests, ISL=8K, OSL=1
- trtllm-bench with concurrency=1

| Backend  | Branch   | max_num_tokens | Req/s | Out tok/s | Total tok/s | Avg Latency (ms) |
|----------|----------|----------------|-------|-----------|-------------|------------------|
| pytorch  | main     | 32768 (no chunking)    | 7.80  | 7.80      | 63929.70    | 128.12           |
| pytorch  | overlap  | 32768 (no chunking)    | 7.81  | 7.81      | 63974.95    | 128.03           |
| pytorch  | main     | 2048 (chunking)        | 6.86  | 6.86      | 56213.56    | 145.71           |
| **pytorch**  | **overlap**  | **2048 (chunking)**        | **9.98**  | **9.98**      | **81730.08**    | **100.21**           |
| tensorrt | main     | 32768 (no chunking)    | 6.97  | 6.97      | 57126.19    | 143.38           |
| tensorrt | overlap  | 32768 (no chunking)    | 7.14  | 7.14      | 58478.51    | 140.07           |
| tensorrt | main     | 2048 (chunking)        | 4.84  | 4.84      | 39644.04    | 206.63           |
| **tensorrt** | **overlap**  | **2048 (chunking)**        | **6.87**  | **6.87**      | **56272.15**    | **145.56**           |

## Test Coverage

- Updated the `llm_get_stats_test_harness` to include chunked prefill and pipeline parallelism support. 
- Added micro batch ID tracking to verify the new pipeline parallel mode behavior with chunked prefill enabled.
- Added test cases for PP size 2 and 4.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

Funatiq committed:
[feat] Overlap context chunks in pipeline parallel mode

- For context chunks there is no dependency on the results of the last pipeline rank, so they can be scheduled in each iteration.
- To achieve this context requests that are chunking are not added to inflight set, so they are scheduled in the next micro batch.
- Context requests that reach the last context chunk are added to inflight set, so they are not scheduled in the next micro batch and generation can run without overlap.
- Enhanced logging for inflight set management.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[fix] Fix max_num_sequences logging in _util.py

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
[feat] Overlap context chunks in pipeline parallel mode

- Added `finished_ctx_reqs` to `BatchStatePP` to track completed context requests.
- Updated `_add_inflight_ids` to return finished context requests for better state management.
- Enhanced `_remove_inflight_ids` to utilize finished context requests from `BatchStatePP`.
- Added debug logging for queuing decisions and inflight request management.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #25072](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25072/) [ run ] completed with state `SUCCESS`. Commit: `4d9fc3c` 
[/LLM/main/L0_MergeRequest_PR pipeline #18952 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/18952/) completed with status: 'SUCCESS'
Pipeline passed with automatic retried tests. Check the [rerun report](https://urm.nvidia.com/artifactory/sw-tensorrt-generic/llm-artifacts//LLM/main/L0_MergeRequest_PR/18952/test-results/rerun_report.html) for details.

tensorrt-cicd commented:
[PR_Github #25221](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25221/) [ run ] completed with state `SUCCESS`. Commit: `4d9fc3c` 
[/LLM/main/L0_MergeRequest_PR pipeline #19076](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19076/) completed with status: 'FAILURE'

Funatiq commented:
/bot run

Funatiq committed:
[chore] Improve documentation for inflight request IDs.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

pcastonguay reviewed (commented):
lgtm. Is there a test we could write to verify the changes work as expected? For example, check the number of iterations required to process the request.

tensorrt-cicd commented:
[PR_Github #25338](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25338/) [ run ] completed with state `SUCCESS`. Commit: `767d19d` 
[/LLM/main/L0_MergeRequest_PR pipeline #19165](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19165/) completed with status: 'FAILURE'

Funatiq committed:
[feat] Enhance LLM stats tests with chunked prefill support

- Updated test parameters to include `enable_chunked_prefill` for both synchronous and asynchronous LLM stats tests.
- Modified `validate_stats` function to account for chunked prefill behavior in result validation.
- Improved test harnesses to handle new parameter and ensure correct behavior with chunked prefill enabled.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

achartier reviewed (approved):
lgtm

Funatiq committed:
[test] Add LLM stats tests for pipeline parallel mode

- Introduced new test cases for LLM stats to validate behavior with multiple pipeline parallel configurations.
- Added micro batch ID tracking to LLM stats and verify it in the test cases.
- Used the new test cases to verify the new pipeline parallel mode behavior with chunked prefill enabled.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #25418](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25418/) [ run ] triggered by Bot. Commit: `f25c023`

tensorrt-cicd commented:
[PR_Github #25502](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25502/) [ run ] triggered by Bot. Commit: `f25c023`

yuxianq commented on tests/unittest/llmapi/test_llm_pytorch.py:
```diff
@@ -54,36 +54,40 @@ def test_tinyllama_logits_processor(enable_chunked_prefill):
-        (False, True, True),
+    "return_context_logits, use_overlap, enable_chunked_prefill, enable_iter_req_stats",
+    [
+        (False, False, False, True),
```
Should we also add `enable_chunked_prefill=True` cases here?

Funatiq committed:
[test] Update LLM stats tests for async behavior

- Added new test cases to validate LLM stats with additional parameters for chunked prefill and overlap.
- Modified the async test harness to pass the repetition index, allowing for better context management during recurrent usage.
- Enhanced validation logic to account for context reuse in subsequent repetitions.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #25703](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/25703/) [ run ] completed with state `SUCCESS`. Commit: `ef031d4` 
[/LLM/main/L0_MergeRequest_PR pipeline #19485](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/19485/) completed with status: 'SUCCESS'
